torchmultimodal/modules/layers/transformer.py

```diff
@@ -29,6 +30,317 @@
+        layer_norm_eps: float = 1e-12,
+        norm_first: bool = False,
+    ) -> None:
+        super().__init__()
+        # attention block
+        self.attention = MultiHeadAttention(
+            dim_q=d_model,
+            dim_kv=d_model,
+            n_head=n_head,
+            attn_module=SelfAttention(dropout),
+        )
+        self.attention_dropout = nn.Dropout(dropout)
+        # feedforward block
+        self.feedforward = MLP(
+            d_model, d_model, dim_feedforward, dropout=dropout, activation=activation
+        )
+        self.feedforward_dropout = nn.Dropout(dropout)
+        # layernorms
+        self.layernorm_first = nn.LayerNorm(d_model, eps=layer_norm_eps)
+        self.layernorm_second = nn.LayerNorm(d_model, eps=layer_norm_eps)
+        self.norm_first = norm_first
+
+    def _attention_block(
+        self,
+        hidden_states: Tensor,
+        attention_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+    ) -> Tuple[Tensor, Tensor]:
+        output, attn_weights = self.attention(
+            hidden_states,
+            attention_mask=attention_mask,
+            head_mask=head_mask,
+            return_attn_weights=True,
+        )
+        output = self.attention_dropout(output)
+        return output, attn_weights
+
+    def _feedforward_block(self, hidden_states: Tensor) -> Tensor:
+        h = self.feedforward(hidden_states)
+        h = self.feedforward_dropout(h)
+        return h
+
+    def _forward_prenorm(
+        self,
+        hidden_states: Tensor,
+        attention_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+        return_attn_weights: bool = False,
+    ) -> Union[Tensor, Tuple[Tensor, Tensor]]:
+        x = hidden_states
+        inputs = _apply_layernorm(x, self.layernorm_first)
+        attn_output, attn_weights = self._attention_block(
+            inputs,
+            attention_mask=attention_mask,
+            head_mask=head_mask,
+        )
+        attn_residual = attn_output + x
+        ff_residual = attn_residual + self._feedforward_block(
+            _apply_layernorm(attn_residual, self.layernorm_second)
+        )
+        if return_attn_weights:
+            return ff_residual, attn_weights
+        else:
+            return ff_residual
+
+    def _forward_postnorm(
+        self,
+        hidden_states: Tensor,
+        attention_mask: Optional[Tensor] = None,
+        head_mask: Optional[Tensor] = None,
+        return_attn_weights: bool = False,
+    ) -> Union[Tensor, Tuple[Tensor, Tensor]]:
+        x = hidden_states
+        attn_output, attn_weights = self._attention_block(
+            x,
+            attention_mask=attention_mask,
+            head_mask=head_mask,
+        )
+        attn_residual = attn_output + x
+        ff_residual = attn_residual + self._feedforward_block(
+            _apply_layernorm(attn_residual, self.layernorm_first)
+        )
```

Comment:
```
attention_norm_output = _apply_layernorm(attn_residual, self.layernorm_first)
ff_residual = attention_norm_output + self._feedforward_block(attention_norm_output)
```
I think the same is true for `_forward_prenorm` but I only tested on `_forward_postnorm`.